// react-browser-extension-scripts/config/paths.js
// @remove-on-eject-begin
/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @remove-on-eject-end
'use strict';

const path = require('path');
const fs = require('fs');

// Make sure any symlinks in the project folder are resolved:
// https://github.com/facebook/create-react-app/issues/637
const appDirectory = fs.realpathSync(process.cwd());
const resolveApp = relativePath => path.resolve(appDirectory, relativePath);

const moduleFileExtensions = [
  'web.mjs',
  'mjs',
  'web.js',
  'js',
  'web.ts',
  'ts',
  'web.tsx',
  'tsx',
  'json',
  'web.jsx',
  'jsx',
];

// Resolve file paths in the same order as webpack
const resolveModule = (resolveFn, filePath) => {
  const extension = moduleFileExtensions.find(extension =>
    fs.existsSync(resolveFn(`${filePath}.${extension}`))
  );

  if (extension) {
    return resolveFn(`${filePath}.${extension}`);
  }

  return resolveFn(`${filePath}.js`);
};

// config after eject: we're in ./config/
module.exports = {
  dotenv: resolveApp('.env'),
  appPath: resolveApp('.'),
  appBuild: resolveApp('build'),
  devAppBuild: resolveApp('dev'),
  appPublic: resolveApp('public'),
  manifestJson: resolveApp('public/manifest.json'),
  appOptionsHtml: resolveApp('public/options.html'),
  appPopupHtml: resolveApp('public/popup.html'),
  appIndexJs: resolveModule(resolveApp, 'src/index'),
  appBackgroundJs: resolveModule(resolveApp, 'src/background/index'),
  appContentScriptJs: resolveModule(resolveApp, 'src/contentScript/index'),
  appOptionsJs: resolveModule(resolveApp, 'src/options/index'),
  appPackageJson: resolveApp('package.json'),
  appSrc: resolveApp('src'),
  appTsConfig: resolveApp('tsconfig.json'),
  appJsConfig: resolveApp('jsconfig.json'),
  yarnLockFile: resolveApp('yarn.lock'),
  testsSetup: resolveModule(resolveApp, 'src/setupTests'),
  appNodeModules: resolveApp('node_modules'),
};

// @remove-on-eject-begin
const resolveOwn = relativePath => path.resolve(__dirname, '..', relativePath);

// config before eject: we're in ./node_modules/react-scripts/config/
module.exports = {
  dotenv: resolveApp('.env'),
  appPath: resolveApp('.'),
  appBuild: resolveApp('build'),
  devAppBuild: resolveApp('dev'),
  appPublic: resolveApp('public'),
  manifestJson: resolveApp('public/manifest.json'),
  appOptionsHtml: resolveApp('public/options.html'),
  appPopupHtml: resolveApp('public/popup.html'),
  appIndexJs: resolveModule(resolveApp, 'src/index'),
  appBackgroundJs: resolveModule(resolveApp, 'src/background/index'),
  appContentScriptJs: resolveModule(resolveApp, 'src/contentScript/index'),
  appOptionsJs: resolveModule(resolveApp, 'src/options/index'),
  appPackageJson: resolveApp('package.json'),
  appSrc: resolveApp('src'),
  appTsConfig: resolveApp('tsconfig.json'),
  appJsConfig: resolveApp('jsconfig.json'),
  yarnLockFile: resolveApp('yarn.lock'),
  testsSetup: resolveModule(resolveApp, 'src/setupTests'),
  appNodeModules: resolveApp('node_modules'),
  // These properties only exist before ejecting:
  ownPath: resolveOwn('.'),
  ownNodeModules: resolveOwn('node_modules'), // This is empty on npm 3
  appTypeDeclarations: resolveApp('src/react-app-env.d.ts'),
  ownTypeDeclarations: resolveOwn('lib/react-app.d.ts'),
};

const ownPackageJson = require('../package.json');
const reactScriptsPath = resolveApp(`node_modules/${ownPackageJson.name}`);
const reactScriptsLinked =
  fs.existsSync(reactScriptsPath) &&
  fs.lstatSync(reactScriptsPath).isSymbolicLink();

// config before publish: we're in ./packages/react-scripts/config/
if (
  !reactScriptsLinked &&
  __dirname.indexOf(path.join('packages', 'react-scripts', 'config')) !== -1
) {
  const templatePath = '../cra-template/template';
  module.exports = {
    dotenv: resolveOwn(`${templatePath}/.env`),
    appPath: resolveApp('.'),
    appBuild: resolveOwn('../../build'),
    devAppBuild: resolveApp('../../dev'),
    appPublic: resolveOwn(`${templatePath}/public`),
    manifestJson: resolveApp(`${templatePath}/public/manifest.json`),
    appPopupHtml: resolveOwn(`${templatePath}/public/popup.html`),
    appOptionsHtml: resolveOwn(`${templatePath}/public/options.html`),
    appIndexJs: resolveModule(resolveOwn, `${templatePath}/src/index`),
    appBackgroundJs: resolveModule(
      resolveOwn,
      `${templatePath}/src/background/index`
    ),
    appContentScriptJs: resolveModule(
      resolveOwn,
      `${templatePath}/src/contentScript/index`
    ),
    appOptionsJs: resolveModule(
      resolveOwn,
      `${templatePath}/src/options/index`
    ),
    appPackageJson: resolveOwn('package.json'),
    appSrc: resolveOwn(`${templatePath}/src`),
    appTsConfig: resolveOwn(`${templatePath}/tsconfig.json`),
    appJsConfig: resolveOwn(`${templatePath}/jsconfig.json`),
    yarnLockFile: resolveOwn(`${templatePath}/yarn.lock`),
    testsSetup: resolveModule(resolveOwn, `${templatePath}/src/setupTests`),
    appNodeModules: resolveOwn('node_modules'),
    // These properties only exist before ejecting:
    ownPath: resolveOwn('.'),
    ownNodeModules: resolveOwn('node_modules'),
    appTypeDeclarations: resolveOwn(`${templatePath}/src/react-app-env.d.ts`),
    ownTypeDeclarations: resolveOwn('lib/react-app.d.ts'),
  };
}
// @remove-on-eject-end

module.exports.moduleFileExtensions = moduleFileExtensions;
